### PR TITLE
geo: implement ST_Summary for Geography/Geometry

### LINKED ISSUE #48405: geo/geogfn: implement ST_Summary({geography})
Implement `ST_Summary` on arguments {geography}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_Summary.html).

For Geography builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geogfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geogfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end.
* Ensure the documentation is regenerated by calling `make buildshort`.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48529 for an example PR.

The following additional guidance has been issued on implementing this function:

Probably doable just by converting to geom.T using the AsGeomT() function and calling various methods.

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2020-05-17T19:28:48Z. Changes to titles, body and labels may be overwritten.</sub>

### PR DESCRIPTION
**Pull request by hueypark**
Fixes #48405, #49049

Release note (sql change): Implemented the geometry based builtins `ST_Summary`.

### CODE REVIEW

File: docs/generated/sql/functions.md [hunk 1]
```diff
@@ -1051,6 +1051,22 @@ has no relationship with the commit order of concurrent transactions.</p>
 </span></td></tr>
 <tr><td><a name="st_startpoint"></a><code>st_startpoint(geometry: geometry) &rarr; geometry</code></td><td><span class="funcdesc"><p>Returns the first point of a geometry which has shape LineString. Returns NULL if the geometry is not a LineString.</p>
 </span></td></tr>
+<tr><td><a name="st_summary"></a><code>st_summary(geography: geography) &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>Returns a text summary of the contents of the geography.</p>
+<p>Flags shown square brackets after the geometry type have the following meaning:
+M: has M coordinate
```
Comment (otan) [inline docs/generated/sql/functions.md:1056]: the generated text here will not look good in HTML.
if you add `*` in front of these in the info string, these should appear as a HTML list, i.e.
```
* M: has M coordinate
* Z: has Z coordinate
* B: has a cached bounding box
* G: is geography
* S: has spatial reference system
```

File: pkg/geo/summary.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+//
+// Flags shown square brackets after the geometry type have the following meaning:
+// M: has M coordinate
+// Z: has Z coordinate
+// B: has a cached bounding box
+// G: is geography
+// S: has spatial reference system
+func Summary(t geom.T, isGeography bool, offset int) (summary string, err error) {
+	shape, err := geopbShape(t)
+	if err != nil {
```
Comment (otan) [inline pkg/geo/summary.go:32]: nit: we could pass in `shape` from `geometry.Shape()` or `geography.Shape()` and avoid rechecking this again.

File: pkg/geo/summary.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+		return "", err
+	}
+
+	f, err := flag(t, isGeography)
```
Comment (otan) [inline pkg/geo/summary.go:36]: nit: flag name conflicts with package `flag` and is a little vague. maybe `summaryFlag`?

File: pkg/geo/summary.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	switch t := t.(type) {
+	case *geom.Point:
+		return summary + fmt.Sprintf("%s[%s]", shape.String(), f), nil
+	case *geom.LineString:
+		return summary + fmt.Sprintf("%s[%s] with %d points", shape.String(), f, t.NumCoords()), nil
+	case *geom.Polygon:
+		numLinearRings := t.NumLinearRings()
+
+		summary += fmt.Sprintf("%s[%s] with %d ring", shape.String(), f, t.NumLinearRings())
+		if 1 < numLinearRings {
```
Comment (otan) [inline pkg/geo/summary.go:51]: nit: easier to read this (and below) as `numLinearRings > 1`

File: pkg/geo/summary.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+		numPoints := t.NumPoints()
+
+		summary += fmt.Sprintf("%s[%s] with %d element", shape.String(), f, numPoints)
+		if 1 < numPoints {
+			summary += "s"
+		}
+
+		for i := 0; i < numPoints; i++ {
+			point := t.Point(i)
+			tmp, err := Summary(point, isGeography, offset+2)
```
Comment (otan) [inline pkg/geo/summary.go:71]: nit: the variable named `tmp` here may be better named as `line`

File: pkg/geo/summary.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+// Summary returns a text summary of the contents of the geometry type.
+//
+// Flags shown square brackets after the geometry type have the following meaning:
+// M: has M coordinate
+// Z: has Z coordinate
+// B: has a cached bounding box
+// G: is geography
+// S: has spatial reference system
+func Summary(t geom.T, isGeography bool, offset int) (summary string, err error) {
```
Comment (otan) [inline pkg/geo/summary.go:30]: nit: taking `offset` as an initial argument for the public facing function makes the interface a little unfriendly as there is a dangling magic constant `0`.

Can we do something like:
```
func Summary(t geom.T, isGeography bool) (summary string, err error) {
   return summary(t, isGeography, 0)
}

func summary(t geom.T, isGeography bool, spaceOffset int) (summary string, err error) {
}
```
```

**Review by otan — COMMENTED** [review]
thanks for coming again and doing this!
few minor things :)
nit: title can just be `geo: implement ST_Summary for Geography/Geometry`

File: docs/generated/sql/functions.md (hunk 1 above)
Comment (hueypark) [reply docs/generated/sql/functions.md]: Done.

File: pkg/geo/summary.go (hunk 2 above)
Comment (hueypark) [reply pkg/geo/summary.go]: Done.

File: pkg/geo/summary.go (hunk 3 above)
Comment (hueypark) [reply pkg/geo/summary.go]: Done.

File: pkg/geo/summary.go (hunk 4 above)
Comment (hueypark) [reply pkg/geo/summary.go]: Done.

File: pkg/geo/summary.go (hunk 5 above)
Comment (hueypark) [reply pkg/geo/summary.go]: Done.

File: pkg/geo/summary.go (hunk 6 above)
Comment (hueypark) [reply pkg/geo/summary.go]: Done.

**Comment by hueypark** [comment]
Thank you for the review!

> nit: title can just be `geo: implement ST_Summary for Geography/Geometry`

Done.

File: pkg/geo/summary.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+			if err != nil {
+				return "", err
+			}
+
+			sum += "\n" + line
+		}
+
+		return sum, nil
+	default:
+		return "", errors.Newf("unspport geom type: %T", t)
```
Comment (otan) [inline pkg/geo/summary.go:142]: i've fixed this typo

File: pkg/geo/summary.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+// M: has M coordinate
+// Z: has Z coordinate
+// B: has a cached bounding box
+// G: is geography
+// S: has spatial reference system
+func Summary(t geom.T, shape geopb.Shape, isGeography bool) (string, error) {
+	return summary(t, shape, isGeography, 0)
+}
+
+func summary(t geom.T, shape geopb.Shape, isGeography bool, offset int) (sum string, err error) {
```
Comment (otan) [inline pkg/geo/summary.go:34]: renamed `summaryLine`

**Review by otan — APPROVED** [review]

**Pull request merged**